include/pmacc/algorithms/reverseBits.hpp

```diff
@@ -43,7 +43,7 @@ T
 reverseBits(T value)
 {
     PMACC_STATIC_ASSERT_MSG( boost::is_integral<T>::value && boost::is_unsigned<T>::value,
-                             Only_allowed_for_unsigned_integral_types );
+                             Only_allowed_for_unsigned_integral_types, );
```

Comment:
Please wait with merge. Let us check if we can workaround it different. The problem I see here is that you need to know that the macro is using `__VA_ARGS__`.

Reply:
I do not think this is a problem: When you use macro you need to know its interface anyway. The distinction is not whether the macro actually uses the `__VA_ARGS__` value, but whether it is a variadic macro, i.e. whether the argument list ends with an ellipsis. There will  only be a trailing comma, if there are no variadic arguments given in the invocation of a variadic macro.

Reply:
We use sometimes macros to give the user a better look and feel. It is not obvious for someone why there mist be a comma after the last argument.

Reply:
I know now how to get it running without destroying the user  side.

```
#define identifier_do(name,...) \
  PMACC_identifier(name,__COUNTER__, __VA_ARGS__  ) 
#define identifier(...) identifier_do(__VA_ARGS__)
```

Reply:
@jkelling Could you please remove the commit with the comma at the end. We will ship a fix in a seperate PR. I will add a intermediate definition everywhere where the user normally must add the comma.

Reply:
STOP wait my test code is wrong

Reply:
After few hours I found a solution to move the pain to the developer instead the users.

```
#include <boost/preprocessor/variadic/size.hpp>
#include <boost/preprocessor/punctuation/comma_if.hpp>
#include <boost/preprocessor/arithmetic/dec.hpp>
#include <boost/preprocessor/comparison/equal.hpp>

#define PMACC_COMMA_IF(count, ...) BOOST_PP_COMMA_IF(BOOST_PP_EQUAL(count,BOOST_PP_VARIADIC_SIZE(__VA_ARGS__)))
#define TAIL(head, ...) __VA_ARGS__

#define PMACC_identifier(name,id,...) \
        struct name \
        { \
                __VA_ARGS__ \
        };

#define identifier_do2(name,...) PMACC_identifier(name,__COUNTER__, __VA_ARGS__ ) 
#define  identifier_do(...) identifier_do2( __VA_ARGS__ )
#define identifier(...) identifier_do(__VA_ARGS__ PMACC_COMMA_IF(1,__VA_ARGS__))

int main()
{
        identifier(xx)
        identifier(zz, int x;)
        identifier(idd, int x;)

        return 0;
}
```

So let us ship the fixin a separate PR.

Reply:
`icc 19`, `clang10` and `gcc 7.4.0` and `gcc 8.2.0` working fine with this solution

Reply:
This does indeed compile without warning.

Are you sure this is worth it? It is a lot of code to add some syntactic sugar on the use side, which if not present only causes a warning.

The PIConGPU core could just lead by example and add the trailing comma, leaving the user to put it there or not. In PRs for new examples, core devolpers could enforce this.

Reply:
To explain: My problem with this being a lot of code is, that this makes is harder for a reader to figure out what these macros are actually doing, putting a higher burden on the documentation, which now needs to explain that
1. only `identifier(...)` is the one to use and what it is for
2. the other are implementation details of `identifier(...)`
3. what the intermediated staps are good for.

Reply:
I do not have a particular opinion. I think the trailing comma would be not expected by some users or developers (me included before this conversation), so that could lead to mysterious and probably hard-to-figure-out build errors. However, I also see your point about adding even more complexity to enable this small thing.

Reply:
@psychocoderHPC you may either need to annouce a descision or put this to a vote for all picongpu developers.

Reply:
Btw.: If you have a trailing comma in an invocation of a non-variadic macro,
```c++
#define foo(name) name;

int main()
{
	foo(int a,)
	return 0;
}
```
 gcc does give a pretty straight error message:
```
foo.cpp:5:12: error: macro "foo" passed 2 arguments, but takes just 1
    5 |  foo(int a,)
      |            ^
foo.cpp:1: note: macro "foo" defined here
    1 | #define foo(name) name;
      | 
foo.cpp: In function ‘int main()’:
foo.cpp:5:2: error: ‘foo’ was not declared in this scope
    5 |  foo(int a,)
      |  ^~~
```
( apart form the last part where `foo` is suddenly not defined :/ , probably because the mactro was not replaced. ).

So, if people get confused and start placing trailing commas in all macro invocation, the compiler will tell them quite clearly that they should not be doing this.

The missing trailing comma is also really only a `-Wpedantic` warning, it is not raised with only `-Wall -Wextra`, so users may not encounter this.

Reply:
@jkelling The parts where we use variadic templates is mostly driven by the fact that we like to have a nice look and feel. `...` is in the macros used to capture only one parameter but if you like to do things like `identifier(foo, Vector<int,5>)` you need to write `identifier(foo, (Vector<int,5>))` if you like to write it correctly and make visible that there are only two parameter for the macro function. To avoid `()` we use `...` in PMacc. Also in some cases we can not use `()` e.g. if create strctures wiht a macro.
The added code will not be the problem but requires a doxygen string to describe the behaviour of `...` for each macro function.

The additional code we need to add is only  two lines
```
#define  identifier_do(...) identifier_do2( __VA_ARGS__ )
#define identifier(...) identifier_do(__VA_ARGS__ PMACC_COMMA_IF(1,__VA_ARGS__))
```

for `identifier()`, `PMACC_ALIGN()` and our compile time asserts.

Reply:
@jkelling could you keep this PR as it is. I will push an update to this PR. I will only change parts the user (PIConGPU developer) is interacting with. There are a few commas which can stay because this is only used internally.